truffle/src/main/ruby/core/exception.rb

```diff
@@ -183,18 +183,18 @@ class LocalJumpError < StandardError
 class NameError < StandardError
 
   def initialize(*args)
-    super(args.shift)
-    Truffle.invoke_primitive :name_error_set_name, self, args.shift
+    name = args.size > 1 ? args.pop : nil
+    super(*args)
+    Truffle.invoke_primitive :name_error_set_name, self, name
   end
 end
 
 class NoMethodError < NameError
-  attr_reader :args
 
   def initialize(*arguments)
-    super(arguments.shift)
-    @name = arguments.shift
-    @args = arguments.shift
+    args = arguments.size > 2 ? arguments.pop : nil
+    super(*arguments) # TODO BJF Jul 24, 2016 Need to handle NoMethodError.new(1,2,3,4)
+    Truffle.invoke_primitive :no_method_error_set_args, self, args
```

Comment:
@eregon I've updated these to match more closely the expected behavior. There is still an edge case here that isn't handled yet mentioned in the `TODO`. It looks like there will be a new field in 2.4 and this edge case behavior is also better documented there.